server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -153,7 +156,7 @@ public RecoveryTarget retryCopy() {
         // If we're retrying we should remove the reference from this instance as the underlying resources
         // get released after the retry copy is created
         Releasable snapshotFileDownloadsPermitCopy = snapshotFileDownloadsPermit;
-        snapshotFileDownloadsPermit = null;
+        snapshotFileDownloadsPermit = SNAPSHOT_FILE_DOWNLOADS_PERMIT_PLACEHOLDER_FOR_RETRY;
```

Comment:
> This commit fixes the problem with a non-null placeholder value indicating that the recovery target was at least originally created while holding some snapshot recovery permits.

Should we add an assertion that snapshotFileDownloadsPermit != null before replacing its value? To assert that the original was holding some snapshot recovery permits as per your comment.

Reply:
Fair point. I don't think it really matters but I made sure we don't go from `null` to non-`null` here in 91ec6cb6a6a8.